### PR TITLE
Neural Modules Configuration stage 2: config import-export

### PR DESCRIPTION
**Pull request by tkornuta-nvidia**
Functionality implemented in this PR:
- [x] Generic export of configuration (init_params) to configuration (yml) file
- [x] Generic import of configuration from configuration file
- [x] Unit tests for both import and export functionalities
- [x] An example script showing how to use import/export
- [x] A tutorial showing how to use import/export
- [x] A tutorial showing how to implement a custom import/export
- [X] Refactor of an existing ASR example to see whether the solution fits to our needs 

The import distinguishes two cases: when importing from the root NeuralModule class and from the leaf neural module.

This PR have also "initializes" the module/config/checkpoint versioning. For now I am saving in config the only available information, i.e. version of the NeMo core plus collection name (collections do not have their own versioning, yet). I am also not performing any version checking during import. Generally I think this issue should be discussed in a separate "Problem brainstorming session".

Additionally, this PR cleans up:
 - [x] utilization of get_module() from NeuralFactory (CV related examples, all relying on ResNet, are left for now - will clean them up along with the introduction of CV collection)

### CODE REVIEW

File: nemo/core/neural_modules.py [hunk 1]
```diff
@@ -162,6 +169,142 @@ def __is_of_allowed_type(self, var):
...
+    def import_config(cls, config_file, config_dir="~/data/configs"):
+        """
+            Class method importing the configuration file.
+            Raises an ImportError exception when config file is invalid or
+            incompatible (when called from a particular class).
+
+            Args:
+                config_file: yml file name
+
+                config_dir: directory where the file will be stored (DEFAULT: ~/data/configs)
```
Comment (redoctopus) [inline nemo/core/neural_modules.py]: will be -> is?

File: nemo/core/neural_modules.py [hunk 2]
```diff
@@ -8,9 +8,11 @@
 from collections import namedtuple
 from enum import Enum
 from inspect import getargvalues, getfullargspec, stack
+from os import path
 from typing import Dict, List, Optional, Set, Tuple
 
-import nemo
+import yaml
```
Comment (blisc) [inline nemo/core/neural_modules.py]: Why yaml? ASR defaults to from ruamel.yaml import YAML

File: nemo/core/neural_modules.py [hunk 3]
```diff
@@ -162,6 +169,142 @@ def __is_of_allowed_type(self, var):
...
+        # Do not check Neural Modules from unit tests.
+        if spec_list[0] == "tests":
+            # Set collection variables.
+            collection_type = "tests"
+            collection_version = None
+        else:
+            # Check if component belongs to any collection
+            if len(spec_list) < 3 or (spec_list[0] != "nemo" and spec_list[1] != "collection"):
+                logging.warning(
+                    "Module `{}` does not belong to any collection. This won't be allowed in the next release.".format(
```
Comment (blisc) [inline nemo/core/neural_modules.py]: Why? What about the common pytorch modules?

File: nemo/core/neural_modules.py [hunk 4]
```diff
@@ -162,6 +169,142 @@ def __is_of_allowed_type(self, var):
...
+        # Add "header" with module "specification".
+        to_export["header"] = {
+            "nemo_core_version": nemo_version,
+            "collection_type": collection_type,
+            "collection_version": collection_version,
+            "class": module_class_name,
+            "full_spec": module_full_spec,
+        }
+        # Add init parameters.
+        to_export["init_params"] = self._init_params
```
Comment (blisc) [inline nemo/core/neural_modules.py]: I might have missed this in the last PR, but does _init_params only support keyword arguments? Or can it support positional arguments as well?

File: requirements/requirements.txt [hunk 5]
```diff
@@ -8,3 +8,4 @@ torch
 torchvision
 wget
 wrapt
+PyYAML
```
Comment (blisc) [inline requirements/requirements.txt]: If PyYAML is now in requirements.txt, please remove it from all other requirements files.

File: nemo/core/neural_modules.py (hunk 2 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_modules.py]: good catch, thx!

File: nemo/core/neural_modules.py (hunk 1 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_modules.py]: of course I have missed that when copy-pasting ;) thx!

File: requirements/requirements.txt (hunk 5 above)
Comment (tkornuta-nvidia) [reply requirements/requirements.txt]: done

File: nemo/core/neural_modules.py (hunk 4 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_modules.py]: You can pass arguments to init() both as positional and keyword arguments

File: nemo/core/neural_modules.py (hunk 3 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_modules.py]: They will be moved to CV collection and Common collection

Comment (tkornuta-nvidia) [reply nemo/core/neural_modules.py]: (not in this PR thou)

File: nemo/core/neural_modules.py (hunk 4 above)
Comment (okuchaiev) [reply nemo/core/neural_modules.py]: should we only allow keyword arguments to simplify everybody's life?

File: nemo/core/neural_modules.py [hunk 6]
```diff
@@ -162,6 +169,142 @@ def __is_of_allowed_type(self, var):
...
+        with open(abs_path_file, 'w') as outfile:
+            yaml.dump(to_export, outfile)
+
+        logging.info(
+            "Configuration of module {} ({}) exported to {}".format(self._uuid, type(self).__name__, abs_path_file)
+        )
+
+    @classmethod
+    def import_config(cls, config_file, config_dir="~/data/configs"):
+        """
```
Comment (okuchaiev) [inline nemo/core/neural_modules.py]: +1 for making it a class method.
The name is a bit confusing, imho, shouldn't this be "import_from_config" or just "from_config" or something else? I had to read the docstring to understand that it returns a NM instance

Comment (tkornuta-nvidia) [reply nemo/core/neural_modules.py]: Originally in the Design Doc I had import_from_config() and export_to_config(), then I simplified it, but yeah, I think you are right, it is less understandable..

File: nemo/core/neural_modules.py (hunk 4 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_modules.py]: yeah, I don't really know... the previous cleanup was _kind of a_ getting rid of positional arguments from all Modules init(). Still, I am not enforcing in any way that a user, when implementing his new module, won't use *args, I am not even sure how to enforce such a thing... :]

File: nemo/core/neural_modules.py (hunk 6 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_modules.py]: done

File: examples/asr/jasper_an4.py [hunk 7]
```diff
@@ -23,61 +22,50 @@
...
-    data_layer = nemo_asr.AudioToTextDataLayer(
-        manifest_filepath=args.train_dataset, labels=vocab, batch_size=args.batch_size, **train_dl_params,
+    # Create a data_layer for training.
+    data_layer = nemo_asr.AudioToTextDataLayer.import_from_config(
+        model_config_file, "AudioToTextDataLayer_train", overwrite_params={"manifest_filepath": args.train_dataset},
     )
 
     num_samples = len(data_layer)
-    steps_per_epoch = math.ceil(num_samples / (args.batch_size * args.iter_per_step * nf.world_size))
+    steps_per_epoch = math.ceil(num_samples / (data_layer.batch_size * args.iter_per_step * nf.world_size))
```
Comment (blisc) [inline examples/asr/jasper_an4.py]: This line is causing jasper_an4 CI test to fail as data_layer.batch_size returns 1, and steps_per_epoch, which is now incorrect, is used to calculate the learning rate policy

Comment (tkornuta-nvidia) [reply examples/asr/jasper_an4.py]: yeah, self._batch_size wasn't set in the AudioDL at all, just pushed a fix, thanks @blisc !

**Comment by okuchaiev** [comment]
looks good to me, what's up with CI though?

File: tests/test_neural_modules_export.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
+# ! /usr/bin/python
```
Comment (blisc) [inline tests/test_neural_modules_export.py]: can you rename this file? you are not testing Neural Module export, you are testing configuration export

File: tests/test_neural_modules_import.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
+# ! /usr/bin/python
```
Comment (blisc) [inline tests/test_neural_modules_import.py]: Can you rename this file too? again configuration import

File: tests/test_neural_modules_export.py (hunk 8 above)
Comment (tkornuta-nvidia) [reply tests/test_neural_modules_export.py]: actually you are right, was thinking about that when renamed examples/tutorial

File: tests/test_neural_modules_import.py (hunk 9 above)
Comment (tkornuta-nvidia) [reply tests/test_neural_modules_import.py]: ok, creating a separate PR
